lib/logstash/outputs/elastic_workplace_search.rb

```diff
@@ -36,72 +40,96 @@ class LogStash::Outputs::ElasticWorkplaceSearch < LogStash::Outputs::Base
-    check_connection!
-  rescue Elastic::WorkplaceSearch::InvalidCredentials
-    raise ::LogStash::ConfigurationError.new("Failed to connect to Workplace Search. Error: 401. Please check your credentials")
-  rescue Elastic::WorkplaceSearch::NonExistentRecord
-    raise ::LogStash::ConfigurationError.new("Failed to connect to Workplace Search. Error: 404. Please check if url '#{@url}' is correct and you've created a source with ID '#{@source}'")
-  rescue => e
-    raise ::LogStash::ConfigurationError.new("Failed to connect to Workplace Search. #{e.message}")
+    @retry_disabled = false
+    @client = LogStash::PluginMixins::EnterpriseSearch::WorkplaceSearch::Client.new(client_options, params: params)
+    begin
+      check_connection!
+    rescue => e
+      raise LogStash::ConfigurationError, "Failed to connect to Workplace Search. Error: #{e.message}"
+    end
   end
 
-  public
   def multi_receive(events)
     # because Workplace Search has a limit of 100 documents per bulk
     events.each_slice(100) do |events|
       batch = format_batch(events)
-      if @logger.trace?
-        @logger.trace("Sending bulk to Workplace Search", :size => batch.size, :data => batch.inspect)
-      end
+      @logger.trace('Sending bulk to Workplace Search', :size => batch.size, :data => batch.inspect) if @logger.trace?
       index(batch)
     end
   end
 
   private
+
+  def client_options
+    options = { :host => @url, :http_auth => @access_token.value }
+    options[:logger] = @logger if @logger.debug?
+    options[:tracer] = @logger if @logger.trace?
+    options
+  end
+
   def format_batch(events)
+    docs_per_source = {}
     events.map do |event|
       doc = event.to_hash
       # we need to remove default fields that start with "@"
       # since Elastic Workplace Search doesn't accept them
       if @timestamp_destination
-        doc[@timestamp_destination] = doc.delete("@timestamp")
+        doc[@timestamp_destination] = doc.delete('@timestamp')
       else # delete it
-        doc.delete("@timestamp")
+        doc.delete('@timestamp')
       end
-      if @document_id
-        doc["id"] = event.sprintf(@document_id)
+
+      doc['id'] = event.sprintf(@document_id) if @document_id
+      doc.delete('@version')
+
+      resolved_source = event.sprintf(@source)
+      unless docs_per_source[resolved_source]
+        @logger.debug('Creating new source segment in batch to send', resolved_source: resolved_source) if @logger.debug?
+        docs_per_source[resolved_source] = []
       end
-      doc.delete("@version")
-      doc
+      docs_per_source[resolved_source] << doc
     end
+
+    docs_per_source
   end
 
-  def index(documents)
-    response = @client.index_documents(@source, documents)
-    report(documents, response)
-  rescue => e
-    @logger.error("Failed to execute index operation. Retrying..", :exception => e.class, :reason => e.message)
-    sleep(1)
-    retry
+  def index(batch)
```

Comment:
```suggestion
  def index(documents_per_source)
```

Could be more explicit on what's a `batch`